design/src/rfcs/rfc0027_new_home_for_cred_cache_and_builder_api_consistency.md

```diff
@@ -0,0 +1,182 @@
+- [x] Add a new `pub` trait `ProvideCachedCredentials` in the `aws-types` crate that has a single method `fn provide_cached_credentials<'a>(&'a self) -> future::ProvideCredentials<'a> where Self: 'a`
+- [x] Update `SharedCredentialsCache` to be a newtype wrapper around `Arc<dyn ProvideCachedCredentials>`
+- [x] Implement `ProvideCachedCredentials` for `SharedCredentialsCache`
+- [x] Implement `ProvideCachedCredentials` for `LazyCredentialsCache`
+
+The reason for the additional changes is that we will need to put a credentials cache into a property bag and get it from the bag in `CredentialsStage`, i.e. in `aws_http::auth` we will have
+```rust
+// formerly known as `set_provider`
+pub fn set_credentials_cache(bag: &mut PropertyBag, cache: SharedCredentialsCache) {
+    bag.insert(cache);
+}
+
+impl CredentialsStage {
+    // --snip--
+
+    async fn load_creds(mut request: Request) -> Result<Request, CredentialsStageError> {
+        let credentials_cache = request
+            .properties()
+            .get::<SharedCredentialsCache>()
+            .cloned();
+
+        // --snip--
+
+        match credentials_cache.provide_cached_credentials().await {
+            Ok(creds) => {
+                request.properties_mut().insert(creds);
+            }
+            // --snip--
+        }
+
+        // --snip--
+    }
+}
+```
+
+Problems
+--------
+Here is how our attempt to implement the selected design in the preceding RFC led to a dead end. Consider this code snippet we are planning to support:
+
+```rust
+let sdk_config = aws_config::from_env()
+    .credentials_cache(CredentialsCache::lazy())
+    .load()
+    .await;
+```
+
+This implies that `aws_config::ConfigLoader::load` will internally have something as follows:
+
+```rust
+pub async fn load(self) -> SdkConfig {
+    // --snip--
+    let credentials_provider = if let Some(provider) = self.credentials_provider {
+        provider
+    } else {
+        // --snip--
+    };
+
+    // We're still massaging the input parameters for `create_cache`
+    // and there could be more parameters, but what's important is
+    // that the method's return type is `SharedCredentialsCache`.
+    let credentials_cache: SharedCredentialsCache =
+        self.credentials_cache.create_cache(credentials_provider);
+
+    // --snip--
+
+    let mut builder = SdkConfig::builder()
+        .region(region)
+        .retry_config(retry_config)
+        .timeout_config(timeout_config)
+        .credentials_cache(credentials_cache) // new method
+        .http_connector(http_connector);
+
+    // --snip--
+
+    builder.build()
+}
+```
+
+The comment `// new method` above further implies that `aws_types::sdk_config::Builder` now has a method `credentials_cache` that takes a `SharedCredentialCache`. What this means is that `aws_types::sdk_config::Builder` has both `credentials_provider` and `credentials_cache` methods. This presents two problems.
```

Comment:
This problem is pretty minor as it goes away if the cache creation takes place inside of the builder rather than the loader. The real issue is the dependency graph.

Reply:
Happy to edit the previous RFC to add a disclaimer for clarity that the builders should be identical all the way through. I'll then just mention the problem of the dependency graph in this PR.

Reply:
Realized that there's no need to edit the previous RFC as it already had the expected contents. Just edited this RFC in 8706c43.